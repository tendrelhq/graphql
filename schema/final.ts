import { resolvers, typeDefs } from "@/schema";
import { makeExecutableSchema, mergeSchemas } from "@graphql-tools/schema";
import { getSchema } from "./v1.schema";

const legacySchema = makeExecutableSchema({ typeDefs, resolvers });
const gratsSchema = getSchema();

export const schema = mergeSchemas({ schemas: [legacySchema, gratsSchema] });
