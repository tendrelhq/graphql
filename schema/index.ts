export type * from "./types";
export type * from "./__generated__/types.generated";
export * from "./__generated__/typeDefs.generated";
export * from "./__generated__/resolvers.generated";
