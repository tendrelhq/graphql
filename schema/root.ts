/** @gqlType */
export type Query = unknown;

/** @gqlType */
export type Mutation = unknown;
