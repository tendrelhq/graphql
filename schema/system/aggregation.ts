/** @gqlType */
export type Aggregate = {
  /**
   * The group, or bucket, that uniquely identifies this aggregate.
   * For example, this will be one of the `groupByTag`s passed to `trackingAgg`.
   *
   * @gqlField
   */
  group: string;

  /**
   * The computed aggregate value.
   *
   * Currently, this will always be a string value representing a duration in
   * seconds, e.g. "360" -> 360 seconds. `null` will be returned when no such
   * aggregate can be computed, e.g. "time in planned downtime" when no "planned
   * downtime" events exist.
   *
   * @gqlField
   */
  value: string | null;
};
