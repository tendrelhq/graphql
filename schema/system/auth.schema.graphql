interface Identity implements Component {
  id: Entity!
}
