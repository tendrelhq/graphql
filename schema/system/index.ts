import z, { type Infer } from "myzod";

const GlobalId = z.string().map(id => {
  const parts = id.split(":");
  if (parts.length !== 2) {
    throw "invariant violated: invalid global identifier";
  }
  return {
    type: parts[0],
    id: parts[1],
  };
});

/**
 * Global Identifier. Externally it just an opaque string. Internally it may
 * have some meaning. Right now this is just the base64 encoding of the
 * object's underlying type (e.g. workinstance) and its uuid (i.e. primary
 * key).
 */
type GlobalId = Infer<typeof GlobalId>;

export function decodeGlobalId(id: unknown): GlobalId {
  if (typeof id !== "string") {
    throw "invariant violated: global ids should be string";
  }
  return GlobalId.parse(
    Buffer.from(decodeURIComponent(id), "base64").toString(),
  );
}

export function encodeGlobalId({ type, id }: GlobalId) {
  return Buffer.from(`${type}:${id}`).toString("base64");
}

// biome-ignore lint/suspicious/noExplicitAny:
export function isGlobalId(value: any): value is GlobalId {
  if (typeof value?.type === "string" && typeof value?.id === "string") {
    return true;
  }

  return false;
}
