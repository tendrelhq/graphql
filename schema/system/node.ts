import { sql } from "@/datasources/postgres";
import { GraphQLError } from "graphql";
import type { ID } from "grats";
import { decodeGlobalId, encodeGlobalId } from ".";
import { Location } from "../platform/archetype/location";
import type { Query } from "../root";
import type { Context } from "../types";
import { Task } from "./component/task";

/**
 * Indicates an object that is "refetchable".
 *
 * @gqlInterface Node
 */
export interface Refetchable {
  /** GraphQL typename. */
  readonly __typename: string;

  /**
   * Internal type, which typically maps to the underlying database table name.
   */
  readonly _type: string;

  /**
   * Internal identifier, which typically maps to the underlying database
   * primary key.
   */
  readonly _id: string;
}

/**
 * A globally unique opaque identifier for a node.
 *
 * @see https://graphql.org/learn/global-object-identification/
 *
 * @gqlField
 * @killsParentOnException
 */
export function id(node: Refetchable): ID {
  return encodeGlobalId({
    type: node._type,
    id: node._id,
  });
}

/**
 * @gqlField
 * @killsParentOnException
 */
export async function node(
  _: Query,
  args: { id: ID },
  ctx: Context,
): Promise<Refetchable> {
  const { type, id } = decodeGlobalId(args.id);
  switch (type) {
    case "location":
      return new Location(args, ctx);
    case "name":
      return {
        __typename: "Name",
        ...(await ctx.orm.name.load(id)),
        // biome-ignore lint/suspicious/noExplicitAny:
      } as any;
    case "organization":
      return {
        __typename: "Organization",
        ...(await ctx.orm.organization.load(id)),
        // biome-ignore lint/suspicious/noExplicitAny:
      } as any;
    case "user":
      return {
        __typename: "User",
        ...(await ctx.orm.user.byId.load(id)),
        // biome-ignore lint/suspicious/noExplicitAny:
      } as any;
    case "workinstance":
    case "worktemplate": {
      const [row] = await sql<[{ type: string }?]>`
        SELECT systagtype AS type
        FROM public.worktemplatetype
        INNER JOIN public.systag
            ON worktemplatetypesystagid = systagid
        WHERE worktemplatetypeworktemplateid IN (
          ${
            type === "workinstance"
              ? sql`SELECT workinstanceworktemplateid FROM public.workinstance WHERE id = ${id}`
              : sql`SELECT worktemplateid FROM public.worktemplate WHERE id = ${id}`
          }
        )
      `;
      switch (row?.type) {
        case "Checklist":
          return {
            __typename: "Checklist",
            id: args.id,
            // biome-ignore lint/suspicious/noExplicitAny:
          } as any;
        default:
          // FIXME: We don't know the parent here!
          return new Task({ id: args.id }, ctx);
      }
    }
    case "workresult":
    case "workresultinstance":
      return {
        __typename: "ChecklistResult",
        id: args.id,
        // biome-ignore lint/suspicious/noExplicitAny:
      } as any;
    case "workpictureinstance":
      // biome-ignore lint/suspicious/noExplicitAny:
      return ctx.orm.attachment.byId.load(args.id) as any;
    default: {
      console.warn(`Unknown node type: '${type}'`);
      throw new GraphQLError("Unknown node type", {
        extensions: {
          code: "BAD_REQUEST",
        },
      });
    }
  }
}
