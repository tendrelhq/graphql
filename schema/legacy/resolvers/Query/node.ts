import type { QueryResolvers } from "@/schema";
import { decodeGlobalId } from "@/schema/system";
import { GraphQLError } from "graphql";

/*
 * Generic "refetch" entrypoint. Useful for ensuring a client has the latest
 * state for a particular "node" (really a query), rather than saying "give me
 * everything that's changed since <some timestamp> though that will be a
 * useful synchronization entrypoint as well.
 */
export const node: NonNullable<QueryResolvers["node"]> = async (
  _,
  args,
  ctx,
) => {
  const { type, id } = decodeGlobalId(args.id);
  switch (type) {
    case "name":
      return {
        __typename: "Name",
        ...(await ctx.orm.name.load(id)),
      };
    case "organization":
      return {
        __typename: "Organization",
        ...(await ctx.orm.organization.load(id)),
      };
    case "user":
      return {
        __typename: "User",
        ...(await ctx.orm.user.byId.load(id)),
      };
    // HACK: Checklists are the only thing that map to these underlying types.
    // In the future we will want to look up the *template types* to determine
    // what their concrete types should be.
    case "workinstance":
    case "worktemplate":
      return {
        __typename: "Checklist",
        id: args.id,
        // biome-ignore lint/suspicious/noExplicitAny:
      } as any;
    case "workresult":
    case "workresultinstance":
      return {
        __typename: "ChecklistResult",
        id: args.id,
        // biome-ignore lint/suspicious/noExplicitAny:
      } as any;
    default:
      throw new GraphQLError("Unknown node type", {
        extensions: {
          code: "BAD_REQUEST",
        },
      });
  }
};
