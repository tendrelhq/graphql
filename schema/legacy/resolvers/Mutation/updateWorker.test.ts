import { describe, expect, test } from "bun:test";
import { randomUUID } from "node:crypto";
import { schema } from "@/schema/final";
import { encodeGlobalId } from "@/schema/system";
import { execute } from "@/test/prelude";
import { SetupTestUpdateWorkerDocument } from "./updateWorker.setup.test.generated";
import { TestUpdateWorkerDocument } from "./updateWorker.test.generated";

describe.skip("updateWorker", () => {
  test("updates worker without conflict", async () => {
    const { data } = await execute(schema, SetupTestUpdateWorkerDocument, {
      input: {
        active: true,
        languageId: "7ebd10ee-5018-4e11-9525-80ab5c6aebee",
        orgId:
          "b3JnYW5pemF0aW9uOmN1c3RvbWVyXzQyY2I5NGVlLWVjMDctNGQzMy04OGVkLTlkNDk2NTllNjhi\nZQ==",
        roleId: "1d8c3097-23f5-4cac-a4c5-ad0a75a181e4",
        firstName: "User",
        lastName: "1337",
      },
    });

    if (!data) throw "failed to setup test";

    const result = await execute(schema, TestUpdateWorkerDocument, {
      input: {
        id: data.createWorker.node.id.toString(),
        languageId: "c3f18dd6-bfc5-4ba5-b3c1-bb09e2a749a9",
        roleId: "a804d5b8-23ef-4592-9486-14857efb1a0a",
        firstName: "1337",
        lastName: "User",
        displayName: "1337 User",
      },
    });

    expect(result.data?.updateWorker.node).toMatchSnapshot();
  });

  test("worker_not_found", async () => {
    const result = await execute(schema, TestUpdateWorkerDocument, {
      input: {
        id: encodeGlobalId({
          type: "workerinstance", // doesn't matter for this test
          id: randomUUID(),
        }),
      },
    });

    expect(result.errors?.at(0)).toMatchObject({
      message: expect.stringMatching(/No Worker for key/),
      extensions: {
        code: "NOT_FOUND",
      },
    });
  });

  test("duplicate_scan_code", async () => {
    const { data } = await execute(schema, SetupTestUpdateWorkerDocument, {
      input: {
        active: true,
        languageId: "7ebd10ee-5018-4e11-9525-80ab5c6aebee",
        orgId:
          "b3JnYW5pemF0aW9uOmN1c3RvbWVyXzQyY2I5NGVlLWVjMDctNGQzMy04OGVkLTlkNDk2NTllNjhi\nZQ==",
        roleId: "1d8c3097-23f5-4cac-a4c5-ad0a75a181e4",
        firstName: "User",
        lastName: "1337",
      },
    });

    if (!data) throw "failed to setup test";

    const result = await execute(schema, TestUpdateWorkerDocument, {
      input: {
        id: data.createWorker.node.id.toString(),
        scanCode: "rugg", // taken by yours truly :D
      },
    });

    expect(result.errors?.at(0)).toMatchObject({
      extensions: {
        code: "duplicate_scan_code",
      },
    });
  });
});
