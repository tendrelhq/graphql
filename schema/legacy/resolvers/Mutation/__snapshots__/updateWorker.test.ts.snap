// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`updateWorker updates worker without conflict 1`] = `
{
  "__typename": "Worker",
  "active": {
    "__typename": "ActivationStatus",
    "active": true,
  },
  "language": {
    "__typename": "Language",
    "code": "es",
  },
  "role": {
    "__typename": "Tag",
    "name": {
      "__typename": "Name",
      "value": "Supervisor",
    },
  },
  "scanCode": null,
  "user": {
    "__typename": "User",
    "displayName": "1337 User",
    "firstName": "1337",
    "lastName": "User",
  },
}
`;
